### PR TITLE
[UP-247] Add GlobalSign DSS example

### PR DESCRIPTION
**Pull request by sampila**

### CODE REVIEW

File: signatures/pdf_sign_external_globalsign.go [hunk 1]
```diff
@@ -0,0 +1,586 @@
...
+	}
+
+	data := make([]byte, sigLen)
+	copy(data, detachedSignature)
+
+	sig.Contents = core.MakeHexString(string(data))
+	return nil
+}
+
+// IsApplicable .
```
Comment (adrg) [inline signatures/pdf_sign_external_globalsign.go:422]: Please complete the comment.

File: signatures/pdf_sign_external_globalsign.go [hunk 2]
```diff
@@ -0,0 +1,586 @@
...
+}
+
+// RevocationInfoArchival is OIDAttributeAdobeRevocation attribute.
+type RevocationInfoArchival struct {
+	Crl          []asn1.RawValue `asn1:"explicit,tag:0,optional"`
+	Ocsp         []asn1.RawValue `asn1:"explicit,tag:1,optional"`
+	OtherRevInfo []asn1.RawValue `asn1:"explicit,tag:2,optional"`
+}
+
+// SignerCallback .
```
Comment (adrg) [inline signatures/pdf_sign_external_globalsign.go:456]: Please complete the comment.

**Review by adrg — COMMENTED** [review]
Looks good overall. I left a couple of minor observations.

File: signatures/pdf_sign_external_globalsign.go [hunk 3]
```diff
@@ -0,0 +1,586 @@
...
+	pdfData, signature, err := getExternalSignatureAndSign(inputPath)
+	if err != nil {
+		log.Fatalf("Fail signature: %v\n", err)
+	}
+
+	// Apply external signature to the PDF data buffer.
+	// Overwrite the generated empty signature with the signature
+	// bytes retrieved from the external service.
+	// Parse signature byte range.
+	byteRange, err := parseByteRange(signature.ByteRange)
```
[the remark this answers is not in the document]
Comment (sampila) [reply signatures/pdf_sign_external_globalsign.go]: Updated at ddbee0a38a3f9965ffe9ba4a532f1aa855f95346

File: signatures/pdf_sign_external_globalsign.go (hunk 2 above)
Comment (sampila) [reply signatures/pdf_sign_external_globalsign.go]: Updated at ddbee0a38a3f9965ffe9ba4a532f1aa855f95346

File: signatures/pdf_sign_external_globalsign.go [hunk 4]
```diff
@@ -0,0 +1,539 @@
...
+	// Write PDF file to buffer.
+	pdfBuf := bytes.NewBuffer(nil)
+	if err = appender.Write(pdfBuf); err != nil {
+		return nil, err
+	}
+
+	return pdfBuf.Bytes(), nil
+}
+
+// getExternalSignature simulates an external service which signs the specified
```
Comment (adrg) [inline signatures/pdf_sign_external_globalsign.go:137]: The name of the function differs from the one mentioned in the comment as well: 
`getExternalSignature` -> `getExternalSignatureAndSign`.
Also, the comment should be changed a bit as the external service is not simulated.

File: signatures/pdf_sign_external_globalsign.go [hunk 5]
```diff
@@ -0,0 +1,539 @@
+/*
+ * This example showcases how to digitally sign a PDF file using an external
+ * signing service which returns PKCS7 package. The external service is
+ * is simulated by signing the file with UniDoc.
```
Comment (adrg) [inline signatures/pdf_sign_external_globalsign.go:4]: The description should be something like:
```
This example showcases how to digitally sign a PDF file using the GlobalSign DSS API.
```

File: signatures/pdf_sign_external_globalsign.go [hunk 6]
```diff
@@ -0,0 +1,539 @@
...
+	}
+	inputPath := args[1]
+	outputPath := args[2]
+
+	apiKey = args[3]
+	apiSecret = args[4]
+	certFilepath = args[5]
+	keyFilepath = args[6]
+
+	// This would be the time to send the PDF buffer to a signing device or
```
Comment (adrg) [inline signatures/pdf_sign_external_globalsign.go:68]: This comment is not really relevant for this example. At least the part of the comment which states that the signature is simulated should be removed as the file is signed using the GlobalSign DSS client.

File: signatures/pdf_sign_external_globalsign.go [hunk 7]
```diff
@@ -0,0 +1,539 @@
...
+	Extra map[string]string
+
+	FilePath string
+
+	// Just in case source file is protected
+	// and defalt password is not empty.
+	Password string
+}
+
+// createSignatureField returns signature field, signature, returns error if fails.
```
Comment (adrg) [inline signatures/pdf_sign_external_globalsign.go:471]: Please rephrase this comment a bit. Perhaps something like:
```
// createSignatureField creates a signature field and an associated signature, 
// based on the specified options.
```

File: signatures/pdf_sign_external_globalsign.go [hunk 8]
```diff
@@ -0,0 +1,539 @@
...
+
+// Sign with custom crypto.Signer which utilize globalsign DSS API.
+func (s *Signer) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) (signature []byte, err error) {
+	if s.callback == nil {
+		return nil, errors.New("signer func not implemented")
+	}
+	return s.callback(rand, digest, opts)
+}
+
+// NewSigner create crypto.Signer implementation
```
Comment (adrg) [inline signatures/pdf_sign_external_globalsign.go:437]: `NewSigner returns a new crypto.Signer implementation.`

File: signatures/pdf_sign_external_globalsign.go [hunk 9]
```diff
@@ -0,0 +1,539 @@
...
+	// Append cert to signature.
+	sig.Cert = pdfCerts
+
+	handler := *es
+	sig.Handler = &handler
+	sig.Filter = core.MakeName("Adobe.PPKLite")
+	sig.SubFilter = core.MakeName("adbe.pkcs7.detached")
+	sig.Reference = nil
+
+	// Reserve initial size
```
Comment (adrg) [inline signatures/pdf_sign_external_globalsign.go:268]: Please remove this comment, as this function actually signs the file.

File: signatures/pdf_sign_external_globalsign.go [hunk 10]
```diff
@@ -0,0 +1,539 @@
...
+	// Annotation font size.
+	FontSize int
+
+	// Extra signature annotation fields.
+	Extra map[string]string
+
+	FilePath string
+
+	// Just in case source file is protected
+	// and defalt password is not empty.
```
[the remark this answers is not in the document]
Comment (sampila) [reply signatures/pdf_sign_external_globalsign.go]: Updated at 15bb7144b66b8891f5c68bd64803bb2d9ee32f54

File: signatures/pdf_sign_external_globalsign.go (hunk 4 above)
Comment (sampila) [reply signatures/pdf_sign_external_globalsign.go]: Updated at 15bb7144b66b8891f5c68bd64803bb2d9ee32f54

File: signatures/pdf_sign_external_globalsign.go (hunk 5 above)
Comment (sampila) [reply signatures/pdf_sign_external_globalsign.go]: Updated at 15bb7144b66b8891f5c68bd64803bb2d9ee32f54

File: signatures/pdf_sign_external_globalsign.go (hunk 6 above)
Comment (sampila) [reply signatures/pdf_sign_external_globalsign.go]: Updated at 15bb7144b66b8891f5c68bd64803bb2d9ee32f54

File: signatures/pdf_sign_external_globalsign.go (hunk 7 above)
Comment (sampila) [reply signatures/pdf_sign_external_globalsign.go]: Updated at 15bb7144b66b8891f5c68bd64803bb2d9ee32f54

File: signatures/pdf_sign_external_globalsign.go (hunk 9 above)
Comment (sampila) [reply signatures/pdf_sign_external_globalsign.go]: Updated at 15bb7144b66b8891f5c68bd64803bb2d9ee32f54

**Review by adrg — APPROVED** [review]
Looks good. Thank you for addressing the review points.

File: signatures/pdf_sign_external_globalsign.go [hunk 11]
```diff
@@ -0,0 +1,586 @@
...
+}
+
+// NewDigest creates a new digest.
+func (es *externalSigner) NewDigest(sig *model.PdfSignature) (model.Hasher, error) {
+	return bytes.NewBuffer(nil), nil
+}
+
+// Validate validates PdfSignature.
+func (es *externalSigner) Validate(sig *model.PdfSignature, digest model.Hasher) (model.SignatureValidationResult, error) {
+	return model.SignatureValidationResult{
```
[the remark this answers is not in the document]
Comment (gunnsth) [reply signatures/pdf_sign_external_globalsign.go:387]: so this is basically only used to fulfill the interface for signing

**Review by gunnsth — APPROVED** [review]
LGTM

**Pull request merged**